FWCore/Concurrency/test/test_catch2_WaitingTaskChain.cc

```diff
@@ -0,0 +1,515 @@
+                   ++count;
+                   REQUIRE(count.load() == 2);
+                 }) |
+                 then([&count](std::exception_ptr const* iPtr, auto h) {
+                   REQUIRE(iPtr == nullptr);
+                   ++count;
+                   REQUIRE(count.load() == 3);
+                 }) |
+                 lastTask(edm::WaitingTaskHolder(group, &waitTask));
+
+        h.doneWaiting(std::exception_ptr());
+      }
+      group.wait();
+      REQUIRE(count.load() == 3);
+      REQUIRE(waitTask.done());
+      REQUIRE(waitTask.exceptionPtr() == nullptr);
+    }
+
+    SECTION("exception -> first | lastTask") {
+      std::atomic<int> count{0};
+
+      edm::FinalWaitingTask waitTask;
+      tbb::task_group group;
+      {
+        using namespace edm::waiting_task::chain;
+        auto h = first([&count](std::exception_ptr const* iPtr, edm::WaitingTaskHolder h) {
+                   REQUIRE(iPtr != nullptr);
+                   ++count;
+                   REQUIRE(count.load() == 1);
+                 }) |
+                 lastTask(edm::WaitingTaskHolder(group, &waitTask));
+
+        h.doneWaiting(std::make_exception_ptr(std::exception()));
+      }
+      group.wait();
+      REQUIRE(count.load() == 1);
+      REQUIRE(waitTask.done());
+      REQUIRE(waitTask.exceptionPtr() == nullptr);
+    }
+
+    SECTION("exception -> first | then | lastTask") {
+      std::atomic<int> count{0};
+
+      edm::FinalWaitingTask waitTask;
+      tbb::task_group group;
+      {
+        using namespace edm::waiting_task::chain;
+        auto h = first([&count](std::exception_ptr const* iPtr, edm::WaitingTaskHolder h) {
+                   REQUIRE(iPtr != nullptr);
+                   ++count;
+                   REQUIRE(count.load() == 1);
+                   h.doneWaiting(*iPtr);
+                 }) |
+                 then([&count](std::exception_ptr const* iPtr, auto h) {
+                   REQUIRE(iPtr != nullptr);
+                   ++count;
+                   REQUIRE(count.load() == 2);
+                 }) |
+                 lastTask(edm::WaitingTaskHolder(group, &waitTask));
+
+        h.doneWaiting(std::make_exception_ptr(std::exception()));
+      }
+      group.wait();
+      REQUIRE(count.load() == 2);
+      REQUIRE(waitTask.done());
+      REQUIRE(waitTask.exceptionPtr() == nullptr);
+    }
+  }
+
+  SECTION("ifException.else testing") {
+    SECTION("first | lastTask") {
+      std::atomic<int> count{0};
+      std::atomic<int> exceptCount{0};
+
+      edm::FinalWaitingTask waitTask;
+      tbb::task_group group;
+      {
+        using namespace edm::waiting_task::chain;
+        auto h = first(ifException([&exceptCount](std::exception_ptr const& iPtr) {
+                         ++exceptCount;
+                         REQUIRE(false);
+                       }).else_([&count](edm::WaitingTaskHolder h) {
+                   ++count;
+                   REQUIRE(count.load() == 1);
```

Comment:
Just to note that the indentation is confusing here (I suppose there is little that can be done for it quickly).